### PR TITLE
Status and SpanContext to struct

### PR DESCRIPTION
**Pull request by lmolkova**
Status and SpanContext are great candidates for readonly struct which helps to reduce allocations.

There is no perf gain with this change that micro-benchmarking can show.

### CODE REVIEW

File: src/OpenTelemetry.Api/Trace/Decision.cs [hunk 1]
```diff
@@ -53,11 +53,13 @@ public Decision(bool isSampled, IEnumerable<KeyValuePair<string, object>> attrib
...
-        public bool IsSampled { get; private set; }
+        public bool IsSampled { get; }
 
         /// <summary>
         /// Gets a map of attributes associated with the sampling decision.
         /// </summary>
-        public IEnumerable<KeyValuePair<string, object>> Attributes { get; private set; }
+        public IEnumerable<KeyValuePair<string, object>> Attributes { get; }
+
+        // TODO ==
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Api/Trace/Decision.cs]: why do we need comparison operator?

File: src/OpenTelemetry.Api/Trace/Status.cs [hunk 2]
```diff
@@ -168,13 +168,21 @@ internal Status(CanonicalCode canonicalCode, string description = null)
...
-            }
-        }
+        public bool IsOk => this.CanonicalCode == CanonicalCode.Ok;
+
+        /// <summary>
+        /// Compare two <see cref="Status"/> for equality.
+        /// </summary>
+        /// <param name="status1">First Status to compare.</param>
+        /// <param name="status2">Second Status to compare.</param>
+        public static bool operator ==(Status status1, Status status2) => status1.Equals(status2);
```
Comment (SergeyKanzhelev) [inline src/OpenTelemetry.Api/Trace/Status.cs]: is it just for a best practice or has some practical use?

File: src/OpenTelemetry.Api/Trace/Decision.cs (hunk 1 above)
Comment (lmolkova) [reply src/OpenTelemetry.Api/Trace/Decision.cs]: it's recommended to override == and != on structs to compare value types

File: src/OpenTelemetry.Api/Trace/Status.cs (hunk 2 above)
Comment (lmolkova) [reply src/OpenTelemetry.Api/Trace/Status.cs]: both, you can't == structs until you override it 

http://csharphelper.com/blog/2017/08/use-and-to-compare-structs-in-c/

**Comment by lmolkova** [comment]
> I'm not clear on the need for == operators. 

To compare structs with ==

> Please implement TODO or remove it before approve

sure, forgot to implement it in decision

**Comment by MikeGoldsmith** [comment]
LGTM.

It's a little confusing #403 and #414's changed were included, which made the review much larger and harder to reason what the change was for. Normally merging from master into a branch wouldn't include those changes in the diff.

**Comment by lmolkova** [comment]
> LGTM.
> 
> It's a little confusing #403 and #414's changed were included, which made the review much larger and harder to reason what the change was for. Normally merging from master into a branch wouldn't include those changes in the diff.

Oops, wasn't intentional. Fill fix that

**Comment by lmolkova** [comment]
> LGTM.
> 
> It's a little confusing #403 and #414's changed were included, which made the review much larger and harder to reason what the change was for. Normally merging from master into a branch wouldn't include those changes in the diff.

Oops, wasn't intentional. Fill fix that

**Comment by MikeGoldsmith** [comment]
Much better, thank you @lmolkova :+1:

**Pull request merged**